Front - basic

andrzejgorski opened the pull request.

andrzejgorski committed:
Add WhylogFront class

Add FrontInput class

Add front input file

andrzejgorski committed:
Add tests for WhylogFront

ppolewicz commented on requirements-test.txt:
```diff
@@ -1,3 +1,4 @@
+nose==1.3.0
+test-generator==0.1.1
+yapf==0.6.2
+mock==1.3.0
```
Why are versions pinned?

ppolewicz commented on whylog/front/__init__.py:
```diff
@@ -0,0 +1,37 @@
+    """
+    This class is the only one that will be used by plugins for text editors
+    implementing Whylog.
+    It provide all funtionalities of whylog.
```
Typos

ppolewicz commented on whylog/front/__init__.py:
```diff
@@ -0,0 +1,37 @@
+        """
+        return self._whylog_teacher.new_teach_object()
+
+    @staticmethod
```
Use classmethod, not staticmethod

ppolewicz commented on whylog/front/__init__.py:
```diff
@@ -0,0 +1,37 @@
+    """
+
+    def __init__(self, **opening_params):
+        self._whylog_client = WhylogClient(**opening_params)
```
_whylog prefix is unnecessary

ppolewicz commented on whylog/front/resource_location.py:
```diff
@@ -0,0 +1,10 @@
+
+
+class AbstractResourceLocation(object):
```
Inheriting class does not need to implement any method? Not even one?

ppolewicz commented on whylog/tests/tests_client/__init__.py:
```diff
@@ -37,7 +37,7 @@ def test_one(self, test_name):
         with open(input_path, 'r') as f:
             vim_line = f.read()
-        result = whylog_client.get_cause(vim_line)
+        # result = whylog_client.get_causes(vim_line)
```
This is wrong. You already know how this line should look like

ppolewicz commented on whylog/tests/tests_front/test_unit_whylog_front.py:
```diff
@@ -0,0 +1,64 @@
+
+
+def get_test_front_input():
+    log_opener = LocallyAccessibleLogOpener(path = '')
```
For tests use InMemoryFile based on StringIO

ppolewicz commented on whylog/tests/tests_front/test_unit_whylog_front.py:
```diff
@@ -0,0 +1,64 @@
+    return FrontInput(
+        offset=1,
+        line_content='effect',
+        resource_location=log_opener
```
Missing comma after last argument

ppolewicz commented on whylog/tests/tests_front/test_unit_whylog_front.py:
```diff
@@ -0,0 +1,64 @@
+
+        assert causes is investigation_result
+
+    def test_get_causes(self):
```
Duplicate method name

ppolewicz commented on whylog/tests/tests_front/test_unit_whylog_front.py:
```diff
@@ -0,0 +1,64 @@
+def get_test_front_input():
+    log_opener = LocallyAccessibleLogOpener(path = '')
+    return FrontInput(
+        offset=1,
```
for such data offset will be `0` (unless the first line of the file is empty, but I doubt this test is about that)

ppolewicz commented on whylog/front/__init__.py:
```diff
@@ -0,0 +1,37 @@
+    @staticmethod
+    def check_teach_object(teach_object, front_input):
+        check_client = WhylogChecker(teach_object)
+        return check_client.get_causes(front_input)
```
I think this should use WhylogClient with passing the teach object into it and then the teach object needs to be passed deep down into the config layer so that it is taken into account as a temporary set of rules and parsers. This will complicate the code somewhat and I think we could maybe skip the requirement of testing the new teach object in the first version - program will still be usable and the implementation will be much easier to understand.

ppolewicz commented on whylog/front/__init__.py:
```diff
@@ -0,0 +1,37 @@
+        return check_client.get_causes(front_input)
+
+    def add_teach_object(self, teach_object):
+        self._whylog_teacher.add_dependence(teach_object)
```
Apparently the caller of WhylogFront will have the teach object already and will just pass it here. It can be only added, not updated or removed, and the only way to get a teach object is to call `new_teach_object()` which takes no parameters. I am not saying this is a good or bad situation, I am just asking if this is what you meant to do.

Could you explain what interface will teach object expose and why is it called a teach object?

ppolewicz commented on whylog/tests/tests_front/__init__.py:
```diff
@@ -0,0 +1,3 @@
+
```
add an empty file rather than one that contains only newlines

andrzejgorski committed:
Refactor

andrzejgorski committed:
Add front utils

andrzejgorski committed:
Rename WhylogClient to Client

andrzejgorski committed:
Refactor get causes name & commented teacher methods

andrzejgorski committed:
Add teacher methods to front

Refactor failed merge

yapf changes

Move const up

isort refactor

pre-commit refactor?

Add Exceptions file

pre-commit another change?

remove unused file

Sorted test-requirements.

Refactor

ppolewicz commented on requirements-test.txt:
```diff
@@ -3,3 +3,5 @@ nose
 pyflakes
 test-generator
 yapf
+mock
```
Please keep this file sorted

ppolewicz commented on whylog/front/__init__.py:
```diff
@@ -1,5 +1,29 @@
+
+    def __init__(self, **opening_params):
+        self._client = Client(**opening_params)
+        self.config = opening_params.get('config', None)
```
Config should not be optional

ppolewicz commented on whylog/front/utils.py:
```diff
@@ -0,0 +1,19 @@
+
+class LocallyAccessibleLogOpener(AbstractResourceLocation):
+    def __init__(self, path):
+        (super(LocallyAccessibleLogOpener, self).__init__(path, LOCALHOST))
```
Outer parenthesis are unnecessary. Hardcoding IP address here doesn't seem to be useful - access will be completely different than in case of remote log, that is why there is a separate class for those two in the first place

ppolewicz commented on whylog/tests/tests_front/test_unit_whylog_front.py:
```diff
@@ -0,0 +1,38 @@
+            mock.return_value = investigation_result
+            causes = self.front.get_causes(front_input)
+
+        assert causes is investigation_result
```
This 6-line test checks a one-line trivial method of Front (`return self.client.get_causes(input_)`). It doesn't bring much value, but it brings test fragility.

ppolewicz commented on whylog/front/utils.py:
```diff
@@ -0,0 +1,19 @@
+LOCALHOST = '127.0.0.1'
```
unused variable

ppolewicz commented on whylog/front/utils.py:
```diff
@@ -0,0 +1,19 @@
+
+
+class AbstractResourceLocation(object):
+    def __init__(self, path, host):
```
some child classes like `LocallyAccessibleLogOpener` don't have an explicitly defined host. Generally resource locations don't need to have a host - they can use service discovery or something like that, to access the log service. I suggest that those two arguments are removed from the interface of `AbstractResourceLocation`.

Also, `AbstractResourceLocation` is not really an abstract class (`@six.add_metaclass(ABCMeta)` is missing)
